README.md
# NinjaVsPirates

My first game, built for android. I can certainly tell this was one of my first projects; always nice to measure progress!

![Ninja Vs Pirates](https://i.imgur.com/T2Qwzm8.png)


changelog.txt
changelog

v1.0.14 - Ad revision, Version Control, Animation and Future Features [BETA]
-----------
- Added a new property to user for ad counter.
- Ads will now display less-frequently. Enjoy :D
- Added tutorial property - This will be used in an upcoming patch for (guess what?) a tutorial.
- Added version property.
- Implemented version control for ease of adding in new properties in the future (thanks Grim).
- Added rotation to ninja stars (delta-time). 
- New property 'kills' added and initial implementation.
- Skins table implemented, more prep for v1.1.

v1.0.13 - Ad removal and final pre-delta-time preperation.
-----------
- Removed ad from onPause event (pause button). Meant to do this earlier. It annoyed me too.
- Changed to 60 FPS - reflected all prior code to this change. Final preperations for full delta-time integration.
  NOTE: This build may perform poorly on older devices. Next build will fix this.
- Removed several pieces of redundant code in place from early delta-time work.
- Reworked structure in several places to eliminate unnecessary function work.
- New FPS function added to run seperately from delta and rounded number.


v1.0.12 - Performance - General Code Cleanup
-----------
- Removed unused forward declare.
- Further optimization changes throughout program in preperation of delta time. Some users may notice a small performance increase.
- Added toggle for display of FPS - Future release will have this in options menu (also not running nearly as fast as it is currently). 

v1.0.11 - Partial delta time support
-----------
- Initial framework laid for delta time
- Changed FPS counter for testing purpose (Yes, it's supposed to update VERY quickly. This is intentional for my own testing.)

v1.0.10 - Animation Edit - Performance Change
-----------
- Changed player hit animation. Should be more noticeable
- Revised java stack size - this may take a performance hit. Testing needed.
- Added FPS meter for beta testing.

v1.0.9 - Menu Addition
-----------
- Added version number on main menu.

v1.0.8 - Final known bugfix and HUGE performance upgrade
-----------
- Converted several files to wav.
- Stereo to mono on several sound effects for optimization.
- Normalized all audio.
- Changed all audio to 22k Hz.
- Fixed audio bug.
- Optimized load and memory storage - should be running and initializing far quicker across all platforms.

v1.0.7 - Performance
-----------
- Enabled largeHeap in build.settings. This should vastly improve performance across all android devices.

v1.0.6 - Beginning of asset enhancements
-----------
- Icon that displays on 'desktop' modified. NEEDS FIXED: https://docs.coronalabs.com/guide/distribution/buildSettings/index.html#androidsettings
- Back-end work on website icons.

v1.0.5 - Performance Boost
-----------
- Additional noticable performance gain by destroying main menu scene upon loading main game scene.
  NOTE: While you will see a slight stutter switching scenes, this allows for a lot more memory to be used in the actual game scene.

v1.0.4 - Various
-----------
- Added instantiated ad to pause button.
- Tweaked what occurs prior to swapping to make transitions smother.
- [N/A but still useful knowledge]: Created poll on twitter (lasting two days). Static or skippable-video ads?

v1.0.3 - Refinement
-----------
- Refined ad network further. Final revision.
- Adjusted poof animation for more tolerence.
- Removed erroneous code that compiled anyhow.
- Removed unnecessary permission.

v1.0.2 - Bugfixes
-----------
- Samsung s5 (possibly others) poof animation glitch corrected.
- Poofs should now properly disappear when two are created simultaneously.

v1.0.1 - Ad change
-----------
- Made ad network a little more dynamic (user feedback positive)

main.lua
-- -----------------------------------------------------------------------------------
-- FILE: main.lua
-- DESCRIPTION: Start the app, declare some globals, and setup the player save file
-- -----------------------------------------------------------------------------------

-- App options
_APPNAME = "Pirates vs Ninjas"
_VERSION = "v1.0.14"
_FONT = "MadeinChina"
_SHOOTUPGRADECOST = 35
_LIVESUPGRADECOST = 100

-- Declare display constants
_CX = display.contentWidth*0.5															-- Center X
_CY = display.contentHeight*0.5															-- Center Y
_CW = display.contentWidth 																-- Width
_CH = display.contentHeight 															-- Height

_T = display.screenOriginY																-- Top
_L = display.screenOriginX																-- Left
_R = display.viewableContentWidth - _L 													-- Right
_B = display.viewableContentHeight - _T 												-- Bottom

-- Hide the status bar
display.setStatusBar( display.HiddenStatusBar )

-- Include composer
local composer = require "composer"

-- Include load/save library
loadsave = require("loadsave")

-- Load audio
_BACKGROUNDMUSIC = audio.loadStream("audio/background-music.mp3")
_THROW = audio.loadSound("audio/throw.wav")
_ENEMYHIT = audio.loadSound("audio/enemy-hit.wav")
_PLAYERHIT = audio.loadSound("audio/player-hit.wav")
_GAMEOVER = audio.loadSound("audio/game-over.wav")
_CLICK = audio.loadSound("audio/click.wav")

-- Set up save file
user = loadsave.loadTable("user.json")
if (user == nil) then
	user = {}
	user.kills = 0
	user.money = 100
	user.adCounter = 1
	user.shootlevel = 0
	user.tutorial = false
	user.version = _VERSION
	user.shootlevelmax = 10
	user.liveslevel = 0
	user.liveslevelmax = 10
	user.playsound = true
	loadsave.saveTable(user, "user.json")
end

-- temporary add-in for existing players. version control. add onto this later.
if (user.version ~= _VERSION or user.version == nil) then
	user.adCounter = 1
	user.version = _VERSION
	user.tutorial = false
	user.kills = 0
	loadsave.saveTable(user, "user.json")
end

skins = loadsave.loadTable("skins.json")
if (skins == nil) then
	skins = {}
	loadsave.saveTable(skins, "skins.json")
end

-- you should have a migrate_save function that accepts palyer info and does stuff to upgrade it to current version

composer.gotoScene("scene_menu")

deltatime.lua
local runtime = 0
 
local function getDeltaTime()
    local temp = system.getTimer()  -- Get current game time in ms
    local dt = (temp-runtime) / (1000/60)  -- 60 fps or 30 fps as base
    runtime = temp  -- Store game time
    return dt
end

-- A box object
local box = display.newRect( 50, 0, 100, 100 )
 
-- Frame update function
local function frameUpdate()
 
   -- Delta Time value
   local dt = getDeltaTime()
 
   -- Move your box, 5 pixels with delta compensation
   box:translate( 0, 5.0*dt )
 
   -- For rotation...
   -- INCORRECT: do not multiply with the entire value!
   -- box.rotation = (box.rotation+1) * dt
   -- CORRECT: only multiply with the changing value:
   box.rotation = box.rotation + (1*dt)
 
   -- Same goes for scaling...
   -- INCORRECT: 1.0 and 0.01 have to be separated, as 1.0 represents the current scale:
   -- box:scale( 1.01*dt, 1.01*dt )
   -- CORRECT: only affect the changing value
   local scale = 1 + (0.01*dt)
   box:scale( scale, scale )
end
 
-- Frame update listener
--Runtime:addEventListener( "enterFrame", frameUpdate )


scene_menu.lua
-- -----------------------------------------------------------------------------------
-- FILE: scene_menu.lua
-- DESCRIPTION: Starts the main menu
-- -----------------------------------------------------------------------------------

local composer = require( "composer" )

local scene = composer.newScene()

local widget = require "widget"
widget.setTheme("widget_theme_ios7")

-- Local forward references
local btn_play, btn_upgrades, btn_sounds
local movePirate, moveNinja

user = loadsave.loadTable("user.json")

-- Switches to scene_game.lua
local function onPlayTouch(event)
    if(event.phase == "ended") then
        audio.play(_CLICK)
        composer.gotoScene("scene_game", "slideLeft")
    end
end

-- Switches to scene_upgrades.lua
local function onUpgradesTouch(event)
    if(event.phase == "ended") then
        audio.play(_CLICK)
        composer.gotoScene("scene_upgrades", "slideUp")
    end
end

-- Toggles sound on and off
local function onSoundsTouch(event)
    if(event.phase == "ended") then
        if(user.playsound == true) then                                                 -- Mute the game
            audio.setVolume(0)
            btn_sounds.alpha = 0.5
            user.playsound = false
        else                                                                            -- Unmute the game
            audio.setVolume(1)
            btn_sounds.alpha = 1
            user.playsound = true
        end
        loadsave.saveTable(user, "user.json")
    end
end

-- -----------------------------------------------------------------------------------
-- Scene creation
-- -----------------------------------------------------------------------------------

-- create()
function scene:create( event )

    local sceneGroup = self.view

    -- Load Graphics
    local background = display.newImageRect(sceneGroup, "images/menuscreen/menu_bg.png", 1425, 950)
    	background.x = _CX; background.y = _CY;

    local backgroundOverlay = display.newImageRect(sceneGroup, "images/menuscreen/menu_bg_overlay.png", 1425, 950)
    	backgroundOverlay.x = _CX; backgroundOverlay.y = _CY;

    local gameTitle = display.newImageRect(sceneGroup, "images/menuscreen/title.png", 508, 210)
        gameTitle.x = _CX; gameTitle.y = _CH * 0.2;

    local myPirate = display.newImageRect(sceneGroup, "images/menuscreen/menu_pirate.png", 209, 358)
        myPirate.x = _L - myPirate.width; myPirate.y = _CH * 0.7; 

    local myNinja = display.newImageRect(sceneGroup, "images/menuscreen/menu_ninja.png", 234, 346)
        myNinja.x = _R + myNinja.width; myNinja.y = _CH * 0.7;

    -- Create some buttons
    local btn_play = widget.newButton {
        width = 426,
        height = 183, 
        defaultFile = "images/menuscreen/btn_play.png", 
        overFile = "images/menuscreen/btn_play_over.png",
        onEvent = onPlayTouch
    }
    btn_play.x = _CX
    btn_play.y = _CY
    sceneGroup:insert(btn_play)

    local btn_upgrades = widget.newButton {
        width = 426,
        height = 183,
        defaultFile = "images/menuscreen/btn_upgrades.png",
        overFile = "images/menuscreen/btn_upgrades_over.png",
        onEvent = onUpgradesTouch
    }
    btn_upgrades.x = _CX
    btn_upgrades.y = btn_play.y + (btn_upgrades.height * 1.25)
    sceneGroup:insert(btn_upgrades)

    btn_sounds = widget.newButton {
        width = 78,
        height = 79,
        defaultFile = "images/menuscreen/btn_music.png",
        overFile = "images/menuscreen/btn_music_over.png",
        onEvent = onSoundsTouch
    }
    btn_sounds.x = _L + 200
    btn_sounds.y = _T + 190
    sceneGroup:insert(btn_sounds)

    -- Version Display (Change version number in main.lua)
    local versionText = display.newText(sceneGroup, _VERSION, 0, 0, _FONT, 50)
        versionText.x = _CX; versionText.y = btn_upgrades.y + (btn_upgrades.height * 0.5 + 25)

    -- Transitions
    local moveNinja = transition.to(myNinja, {x=950, delay = 250})
    local movePirate = transition.to(myPirate, {x=250, delay = 250})

end


-- show()
function scene:show( event )

    local sceneGroup = self.view
    local phase = event.phase

    if ( phase == "will" ) then
        -- Code here runs when the scene is still off screen (but is about to come on screen)

    elseif ( phase == "did" ) then
        -- Code here runs when the scene is entirely on screen
        audio.play(_BACKGROUNDMUSIC, {loops=-1, delay=1500})

        -- Kills previous scene to ensure reloading.
        local prevScene = composer.getSceneName("previous")
        if(prevScene) then 
            composer.removeScene(prevScene)
        end
    end
end


-- hide()
function scene:hide( event )

    local sceneGroup = self.view
    local phase = event.phase

    if ( phase == "will" ) then
        -- Code here runs when the scene is on screen (but is about to go off screen)

    elseif ( phase == "did" ) then
        -- Code here runs immediately after the scene goes entirely off screen

    end
end


-- destroy()
function scene:destroy( event )

    local sceneGroup = self.view
    -- Code here runs prior to the removal of scene's view

end


-- -----------------------------------------------------------------------------------
-- Scene event function listeners
-- -----------------------------------------------------------------------------------
scene:addEventListener( "create", scene )
scene:addEventListener( "show", scene )
scene:addEventListener( "hide", scene )
scene:addEventListener( "destroy", scene )
-- -----------------------------------------------------------------------------------

return scene

scene_upgrades.lua
-- -----------------------------------------------------------------------------------
-- FILE: scene_upgrades.lua
-- DESCRIPTION : Handles the upgrade display view / buttons / actions
-- -----------------------------------------------------------------------------------

local composer = require( "composer" )

local scene = composer.newScene()

local widget = require "widget"
widget.setTheme("widget_theme_ios7")

user = loadsave.loadTable("user.json")

local btn_upgrade1, btn_upgrade2, btn_menu, sceneTitle, warningMessage                  -- Local forward declares

-- -----------------------------------------------------------------------------------
-- Scene functions
-- -----------------------------------------------------------------------------------

-- Goes back to main menu
local function onMenuTouch(event)
    if(event.phase == "ended") then
        audio.play(_CLICK)
        composer.gotoScene("scene_menu", "slideDown")
    end
end

local function hideWarningMessage()
    warningMessage.alpha = 0
end

-- Upgrades fire rate
local function onUpgrade1Touch(event)
    if(event.phase == "ended") then
        audio.play(_CLICK)
        if(user.money >= (user.shootlevel * _SHOOTUPGRADECOST + _SHOOTUPGRADECOST)) then
            -- Allow player to proceed with upgrade
            if(user.shootlevel >= user.shootlevelmax) then                              -- Stops and sets label if max level
                btn_upgrade1:setLabel("Max Level")                                                            
            else                                                                        -- Proceed with upgrade
                user.money = user.money - (user.shootlevel * _SHOOTUPGRADECOST + _SHOOTUPGRADECOST)

                sceneTitle.text = "Upgrades - $"..user.money

                user.shootlevel = user.shootlevel + 1
                loadsave.saveTable(user, "user.json")
                user = loadsave.loadTable("user.json")

                btn_upgrade1:setLabel("$"..(user.shootlevel * _SHOOTUPGRADECOST + _SHOOTUPGRADECOST).."  Rank "..user.shootlevel)
            end
        else                                                                            -- Otherwise, display the warning message of not enough money
            warningMessage.alpha = 1                                            
            local tmr_hidewarningmessage = timer.performWithDelay(750, hideWarningMessage, 1)
        end
    end
end

-- Upgrades lives
local function onUpgrade2Touch(event)
    if(event.phase == "ended") then
        audio.play(_CLICK)
        if(user.money >= (user.liveslevel * _LIVESUPGRADECOST + _LIVESUPGRADECOST)) then                         
            -- Allow player to proceed with upgrade - has enough money
            if(user.liveslevel >= user.liveslevelmax) then                              -- Stops and sets label if max level
                btn_upgrade2:setLabel("Max Level")
            else                                                                        -- Proceed with upgrade
                user.money = user.money - (user.liveslevel * _LIVESUPGRADECOST + _LIVESUPGRADECOST)

                sceneTitle.text = "Upgrades - $"..user.money

                user.liveslevel = user.liveslevel + 1
                loadsave.saveTable(user, "user.json")
                user = loadsave.loadTable("user.json")

                btn_upgrade2:setLabel("$"..(user.liveslevel * _LIVESUPGRADECOST + _LIVESUPGRADECOST).."  Rank "..user.liveslevel)
            end
        else                                                                            -- Otherwise, display the warning message of not enough money            
            warningMessage.alpha = 1
            local tmr_hidewarningmessage = timer.performWithDelay(750, hideWarningMessage, 1)
        end
    end
end


-- create()
function scene:create( event )

    local sceneGroup = self.view
    
-- -----------------------------------------------------------------------------------
-- Scene Creation
-- -----------------------------------------------------------------------------------
    
    -- Load graphics
    local background = display.newImageRect(sceneGroup, "images/menuscreen/menu_bg.png", 1425, 950)
        background.x = _CX; background.y = _CY
    
    local backgroundOverlay = display.newImageRect(sceneGroup, "images/menuscreen/menu_bg_overlay.png", 1425, 950)
        backgroundOverlay.x = _CX; backgroundOverlay.y = _CY
    
    local banner = display.newImageRect(sceneGroup, "images/menuscreen/banner.png", 1250, 200)
        banner.x = _CX; banner.y = _CH * 0.2

    -- Set Text on graphics
    sceneTitle = display.newText(sceneGroup, "Upgrades - $"..user.money, 0, 0, _FONT, 60)
        sceneTitle.x = _CX; sceneTitle.y = banner.y

    warningMessage = display.newText(sceneGroup, "Not enough money.", 0, 0, _FONT, 52)
        warningMessage.x = _CX; warningMessage.y = sceneTitle.y + (sceneTitle.height * 1)
        warningMessage.alpha = 0

    -- Buttons
    btn_upgrade1 = widget.newButton {
        width = 480,
        height = 183, 
        defaultFile = "images/menuscreen/btn_shoot.png", 
        overFile = "images/menuscreen/btn_shoot_over.png",
        font = _FONT,
        fontSize = 60,
        labelColor = {default={1,1,1},over={0,0,0}},
        labelYOffset = 15,
        label = "$"..(user.shootlevel*_SHOOTUPGRADECOST + _SHOOTUPGRADECOST).."  Rank "..user.shootlevel,
        onEvent = onUpgrade1Touch
    }
    btn_upgrade1.x = _CX - (btn_upgrade1.width * 0.6)
    btn_upgrade1.y = _CY
    sceneGroup:insert(btn_upgrade1)

    btn_upgrade2 = widget.newButton {
        width = 480,
        height = 183, 
        defaultFile = "images/menuscreen/btn_lives.png", 
        overFile = "images/menuscreen/btn_lives_over.png",
        font = _FONT,
        fontSize = 60,
        labelColor = {default={1,1,1},over={0,0,0}},
        labelYOffset = 15,
        label = "$"..(user.liveslevel*_LIVESUPGRADECOST + _LIVESUPGRADECOST).."  Rank "..user.liveslevel,
        onEvent = onUpgrade2Touch
    }
    btn_upgrade2.x = _CX + (btn_upgrade2.width * 0.6)
    btn_upgrade2.y = _CY
    sceneGroup:insert(btn_upgrade2)

    local btn_menu = widget.newButton {
        width = 426,
        height = 183, 
        defaultFile = "images/menuscreen/btn_menu.png", 
        overFile = "images/menuscreen/btn_menu_over.png",
        onEvent = onMenuTouch
    }
    btn_menu.x = _CX
    btn_menu.y = btn_upgrade1.y + 260
    sceneGroup:insert(btn_menu)

    -- These values are in main.lua
    if(user.shootlevel >= user.shootlevelmax) then
        btn_upgrade1:setLabel("Max Level")
    end
    if(user.liveslevel >= user.liveslevelmax) then
        btn_upgrade2:setLabel("Max Level")
    end

end


-- show()
function scene:show( event )

    local sceneGroup = self.view
    local phase = event.phase

    if ( phase == "will" ) then
        -- Code here runs when the scene is still off screen (but is about to come on screen)

    elseif ( phase == "did" ) then
        -- Code here runs when the scene is entirely on screen

    end
end


-- hide()
function scene:hide( event )

    local sceneGroup = self.view
    local phase = event.phase

    if ( phase == "will" ) then
        -- Code here runs when the scene is on screen (but is about to go off screen)

    elseif ( phase == "did" ) then
        -- Code here runs immediately after the scene goes entirely off screen

    end
end


-- destroy()
function scene:destroy( event )

    local sceneGroup = self.view
    -- Code here runs prior to the removal of scene's view

end


-- -----------------------------------------------------------------------------------
-- Scene event function listeners
-- -----------------------------------------------------------------------------------
scene:addEventListener( "create", scene )
scene:addEventListener( "show", scene )
scene:addEventListener( "hide", scene )
scene:addEventListener( "destroy", scene )
-- -----------------------------------------------------------------------------------

return scene